docs/interface.md

```diff
@@ -0,0 +1,56 @@
+# The Interface
+
+All interactions with the Exercism website are handled automatically. Test runners have the single responsibility of taking a solution, running its tests and returning a standardised output.
+
+## Execution
+
+- A test runner should provide an executable script. You can find more information in the [docker.md](https://github.com/exercism/automated-mentoring-support/blob/master/docs/docker.md) file.
+- The script will receive two parameters:
+  - The slug of the exercise (e.g. `two-fer`).
+  - A path to a directory containing the submitted file(s) (with a trailing slash).
+- The script must write a file to that directory named `report.json`
+
+## Output format
+
+The `report.json` file should be structured as followed:
+
+```json
+{
+  "status": "fail",
+  "tests": [
+    {
+      "name": "Test that the thing works",
+      "status": "fail",
+      "expected": "42",
+      "actual": "123123",
+      "message": "Expected 42 but got 123123"
+    }
+  ]
+}
+```
+
+The following overall statuses are valid:
+- `pass`: All tests passed
+- `fail`: At least one test failed
+- `error`: To be used when the tests did not run correctly (including if there are any skips)
```

Comment:
This list of overall statuses is incomplete, because it ignores the possibility of optional tests.

The Rust track has several optional, feature-gated tests. Mentors are supposed to approve a student solution even if these tests are not attempted, but encourage students to at least try them out. 

```sh
coriolinus@brute /mnt/c/Users/coriolinus/Documents/Projects/exercism/xrust
$ rg 'cfg\(feature'
Wed Aug 28 19:25:04 DST 2019
exercises/binary-search/tests/binary-search.rs
88:#[cfg(feature = "generic")]
95:#[cfg(feature = "generic")]
104:#[cfg(feature = "generic")]

exercises/doubly-linked-list/src/pre_implemented.rs
62:#[cfg(feature = "advanced")]
71:#[cfg(feature = "advanced")]

exercises/doubly-linked-list/tests/doubly-linked-list.rs
267:#[cfg(feature = "advanced")]
278:#[cfg(feature = "advanced")]

exercises/reverse-string/tests/reverse-string.rs
58:#[cfg(feature = "grapheme")]

exercises/triangle/tests/triangle.rs
158:#[cfg(feature = "generic")]
169:#[cfg(feature = "generic")]
179:#[cfg(feature = "generic")]

exercises/wordy/tests/wordy.rs
165:#[cfg(feature="exponentials")]
173:#[cfg(feature="exponentials")]
```

In order to support this workflow, the test suite must pass if all core tests pass, even if the feature tests fail. 

I propose adding the following status to the overall results: `core_pass`. This is returned when all core tests pass, even though one or more feature-gated tests failed.

Reply:
The junit equivalent is called `skip` for a single test, and I don't think `core_pass` makes much sense. From a running standpoint, both `pass` and "core-pass" would return exit status code 0, so they're the same from a CI standpoint. I can see that a test suite for a track implements two test-sets, one for each possible implementation. In this case you would have to pick _either_ test. Are we going to call that "branch_a-pass" and "branch_b-pass"?

Adding additional "top-level" messaging is common: x tests passed, y assertions passed, z tests skipped, q tests errored, u tests failed. 

Similarly: "All tests passed, 3 optional tests skipped".

Reply:
The test runner can't know whether or not the student has attempted the optional tests. Therefore, it has to run all tests. Therefore, no tests will ever be skipped.

Adding top-level messaging doesn't help: the whole point of having a unified `status` field is that this is the absolute top-level result. If we have this field, then we need to be able to distinguish between "all tests passed" and "all core tests passed, but some optional tests failed."

I would also accept removing the `status` field entirely in favor of `core_failures` and `optional_failures`.

Your hypothetical about implementing two test suites is a strawman; I don't know of any track which actually currently does such a thing. However, Rust has optional tests right now; they are not a separate branch, but a superset of the core tests.

Reply:
> The Rust track has several optional, feature-gated tests. 

Interesting feature! Are there any other languages that have a similar feature? So far, the languages that I've done (except for Rust) seem to all use something like a "Skip" marker.

> In order to support this workflow, the test suite must pass if all core tests pass, even if the feature tests fail.

Could we "solve" this by having the test runner never run the optional tests? That way, only the essential, non-optional tests will be run, and we can determine if all non-optional tests have passed.

Reply:
I don't know of any other language tracks with optional tests like these. 

The problem with using the `#[ignore]` marker on the tests is that we're already using that, with different semantics: to encourage TDD, all but the first test are ignored when students first download the file; as their code passes all tests currently run, they're supposed to iteratively remove the marker from subsequent tests. I'd had the impression that that setup was Exercism standard; it dates from before I started the Rust track as a student, let alone as a maintainer. 

> Could we "solve" this by having the test runner never run the optional tests?

Yes, of course; that's the default behavior. However, I think it's a pretty suboptimal solution: it means that for nontrivial exercises, the mentors still have to download and test student solutions themselves; they can't just review the code and the test result dashboard. I'd rather just build in appropriate support.

Reply:
I'll have a think from a product POV about how we want optional tests to function then reply more here. I think (?) we all agree the statuses here are at least a subset of the final set so I don't feel that this discussion blocks people from making a start, although we might need to extend the set as per @coriolinus' comments.

Reply:
Sounds good!